Fix #1795: Essentials of Scala.js-defined JS classes.

sjrd opened the pull request:
This PR implements most of the specification for Scala.js-defined JavaScript classes, as discussed in #1795.

The following features are not yet supported, and are reported as implementation restrictions:

* #1811 Secondary constructors
* #1812 Exporting a Scala.js-defined JS class
* #1813 Calling the super constructor with `:_*` to a varargs param.

This PR does not implement the deprecations that would lead to nicer definitions either. These aspects will come as follow ups, since they are not critical to the general specification.

sjrd commented:
CI only for now.

Or for the curious, you can dive in. One interesting thing is the tests :)

sjrd commented:
Also, suggestions for more pos tests are welcome :)

gzm0 commented on compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala:
```diff
@@ -878,36 +1030,41 @@ abstract class GenJSCode extends plugins.PluginComponent
+              val genInitialThis = genThis() // outside of withScopedVars
+              val thisSym = thisDef.symbol
+              withScopedVars(
+                thisLocalVarIdent := Some(encodeLocalSym(thisSym))
```
Consider making a `val` out of `encodeLocalSym`s result.

gzm0 replied:
(will allow you to not write `.get.get` below).

sjrd replied:
OK.

gzm0 commented on compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala:
```diff
@@ -878,36 +1030,41 @@ abstract class GenJSCode extends plugins.PluginComponent
+                js.Block(
+                    js.VarDef(thisLocalVarIdent.get.get,
+                        currentClassType, thisSym.isMutable,
+                        genInitialThis),
```
Consider moving all of the above out of the `withScopedVars` block.

sjrd replied:
Indeed, much nicer.

gzm0 commented:
What about overload of normal methods? I didn't see any tests for it.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1963/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1963/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2543/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2543/</a>
Test FAILed.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1964/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1964/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2544/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2544/</a>
Test FAILed.

sjrd commented:
> What about overload of normal methods? I didn't see any tests for it.

https://github.com/scala-js/scala-js/pull/1809/files#diff-6e8c360252085a4ec6f673609bd75631R366 ("simple overloaded methods" + "renamed overloaded methods")

I sort of trust the tests in `ExportsTest` for all the weird combinations or param count/types/repeated/default, since the same code is used in both cases.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1966/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1966/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2547/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2547/</a>
Test FAILed.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1967/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1967/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2548/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2548/</a>
Test FAILed.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1970/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1970/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2551/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2551/</a>
Test FAILed.

sjrd commented:
And there we go! :D
Review by @gzm0, please :-p

More pos tests still welcome, I have this feeling that I must have missed some.

sjrd commented:
Self-review: I forgot to forbid overloading of private methods.

sjrd commented:
The compiler test `ScalaJSDefinedTest.noSuperConstructorCallWithSpread` has different error messages in 2.10 (wrt. positions).

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1971/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1971/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2552/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2552/</a>
Test FAILed.

sjrd commented:
Updated with the error messages to prevent overloading of private methods. And also to prevent over*riding* of private methods, actually.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1974/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1974/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2556/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2556/</a>
Test FAILed.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1976/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1976/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2558/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2558/</a>
Test FAILed.

sjrd commented:
retest this please

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1980/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1980/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2562/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2562/</a>
Test FAILed.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1983/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1983/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2566/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2566/</a>
Test FAILed.

gzm0 commented on compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala:
```diff
@@ -1142,8 +1347,8 @@ abstract class GenJSCode extends plugins.PluginComponent
-      if (methodTailJumpThisSym.get != NoSymbol) {
-        js.VarRef(encodeLocalSym(methodTailJumpThisSym))(currentClassType)
+      if (thisLocalVarIdent.get.isDefined) {
+        js.VarRef(thisLocalVarIdent.get.get)(currentClassType)
```
Still use `fold`?

sjrd replied:
Good point.

gzm0 commented on compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala:
```diff
@@ -878,36 +1030,48 @@ abstract class GenJSCode extends plugins.PluginComponent
+              val thisLocalVarDef = {
+                js.VarDef(thisLocalIdent,
+                    currentClassType, thisSym.isMutable,
+                    genThis())
```
I think you can do this in two lines.

gzm0 replied:
Or event better, without the braces.

sjrd replied:
ok

gzm0 commented on compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala:
```diff
@@ -878,36 +1030,48 @@ abstract class GenJSCode extends plugins.PluginComponent
+            js.Block(otherStats.map(genStat) :+ (
                 if (bodyIsStat) genStat(rhs)
                 else            genExpr(rhs)))
+          }
```
Consider putting the `withScopedVars` in here. It will improve two things:

- The comment is obsolete and it's explicit from the code
- Less code below

sjrd replied:
Should I then give the `Assignment[_]*` as parameter to `genInnerBody()`? Seems weird. I'd rather `withScopedVars` and its `Assignment`s stick together.

gzm0 commented:
LGTM unti bca957a

gzm0 commented:
That's all until 5dc9d6b

gzm0 commented on compiler/src/main/scala/org/scalajs/core/compiler/PrepJSInterop.scala:
```diff
@@ -77,13 +79,29 @@ abstract class PrepJSInterop extends plugins.PluginComponent
+      val result = body
+      enclosingOwner = oldEnclosingOwner
+      allEnclosingOwners = oldAllEnclosingOwners
+      result
```
Hmmm.. maybe we should protect this by a finally block.

sjrd replied:
ok

gzm0 commented on compiler/src/main/scala/org/scalajs/core/compiler/PrepJSInterop.scala:
```diff
@@ -322,6 +351,17 @@ abstract class PrepJSInterop extends plugins.PluginComponent
+        sym.addAnnotation(ScalaJSDefinedAnnotation)
+
+      val isJSNative =
+        !sym.hasAnnotation(ScalaJSDefinedAnnotation)
```
I wouldn't break here..

sjrd replied:
ok

gzm0 commented:
Would you mind separating the state cleanup in `PrepJSInterop` into a separate commit? It is very hard to see which changes come from the Scala.js defined classes implementation and which come from the refactor.

sjrd commented:
> Would you mind separating the state cleanup in PrepJSInterop into a separate commit?

Ah OK I'll do that.

sjrd commented:
Updated

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1985/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1985/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2568/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2568/</a>
Test PASSed.

sjrd commented on compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala:
```diff
@@ -1664,7 +1883,10 @@ abstract class GenJSCode extends plugins.PluginComponent
       } else if (isRawJSType(receiver.tpe) && sym.owner != ObjectClass) {
-        genPrimitiveJSCall(tree, isStat)
+        if (!isScalaJSDefinedJSClass(sym.owner) || jsInterop.isExposed(sym))
+          genPrimitiveJSCall(tree, isStat)
```
Arg! This is completely wrong! It won't work under separate compilation (if the `sym.owner` class was compiled in a separate run), because `jsInterop.isExposed` only works in the same run. How stupid could I be!?

I'll need to use an annotation instead of a `mutable.Set[Symbol]`.

gzm0 replied:
Don't forget about the restrictions we found w.r.t. ignoring invalid descendants.

sjrd replied:
Hum ... there is no such thing as an invalid descendent for Scala.js-defined JS classes. All descendents must be valid. I must be missing something here?

sjrd replied:
Fixed by having `jsInterop.{isExposed,registerJSExposedSymbol}` manipulate annotations under the hood.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1986/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1986/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2569/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2569/</a>
Test FAILed.

sjrd commented:
Updated.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1988/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1988/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2572/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2572/</a>
Test FAILed.

sjrd commented:
Updated. I have added an additional commit which enables Scala.js-defined JS traits if they are pure interfaces, i.e., if they have no concrete term member. This commit could be squashed with the Essentials, probably.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1990/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1990/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2574/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2574/</a>
Test PASSed.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1991/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1991/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2575/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2575/</a>
Test PASSed.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1993/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/1993/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2577/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2577/</a>
Test PASSed.

gzm0 commented:
LGTM until ef31d1a2a7b84509d04bfe54fddde23932ebcc5c

gzm0 commented:
There is a typo in "Refactor PrepJSInterop: clean up the owner chain handling.":

This was hard to read, and had the limitation that we could not tell information **about** the directly enclosing owner, which made some checks even less readable.

gzm0 commented on compiler/src/main/scala/org/scalajs/core/compiler/PrepJSInterop.scala:
```diff
@@ -713,3 +819,46 @@ abstract class PrepJSInterop extends plugins.PluginComponent
+    val JSClass = new OwnerKind(0x10)
+    val EnumClass = new OwnerKind(0x20)
+    val EnumMod = new OwnerKind(0x40)
+    val EnumImpl = new OwnerKind(0x80)
```
Please add short doc comments to the three enum constants. It is hard to distinguish what they mean without looking at how they are used.

gzm0 replied:
You also might want to drop a quick comment that these values form a partition of all possible kinds.

gzm0 commented on compiler/src/main/scala/org/scalajs/core/compiler/PrepJSInterop.scala:
```diff
@@ -109,42 +129,53 @@ abstract class PrepJSInterop extends plugins.PluginComponent
+        enterOwner(OwnerKind.EnumClass) { super.transform(cldef) }
       case idef: ImplDef if isScalaEnum(idef) =>
-        enterScalaEnum { super.transform(idef) }
+        enterOwner(OwnerKind.EnumMod) { super.transform(idef) }
```
Consider pulling these two cases in a single case (and have a short `if` to separate `EnumMod` from `EnumClass`). Or at least change the lower match from `ImplDef` to `ModuleDef`.

gzm0 commented on compiler/src/main/scala/org/scalajs/core/compiler/PrepJSInterop.scala:
```diff
@@ -157,45 +188,62 @@ abstract class PrepJSInterop extends plugins.PluginComponent
       // Catch ValDefs in enumerations with simple calls to Value
-      case ValDef(mods, name, tpt, ScalaEnumValue.NoName(optPar)) if inScalaEnum =>
+      case ValDef(mods, name, tpt, ScalaEnumValue.NoName(optPar))
+          if enclosingOwner is OwnerKind.Enum =>
```
This is strictly speaking a regression:

```scala
class A extends Enumeration {
  val A = Value
  class B {
    val B = Value
  }
}
```

will not be translated like before.

sjrd replied:
OK I will revert to the old behavior with `anyEnclosingOwner is OwnerKind.Enum`.

gzm0 commented:
That's all until "Refactor PrepJSInterop: clean up the owner chain handling".

gzm0 commented on compiler/src/main/scala/org/scalajs/core/compiler/PrepJSInterop.scala:
```diff
@@ -334,29 +375,42 @@ abstract class PrepJSInterop extends plugins.PluginComponent
+        case _ if !isJSNative && !sym.isTrait &&
+            sym.info.parents.exists(_ =:= AnyRefClass.tpe) =>
+          reporter.error(implDef.pos,
+              "A Scala.js-defined JS class cannot extend directly AnyRef. " +
```
cannot directly extend

gzm0 commented on compiler/src/main/scala/org/scalajs/core/compiler/PrepJSInterop.scala:
```diff
@@ -334,29 +375,42 @@ abstract class PrepJSInterop extends plugins.PluginComponent
+        case _ if !isJSNative && sym.info.parents.exists(isNativeJSTraitType) =>
+          val obj = if (sym.isTrait) "trait" else "class"
+          reporter.error(implDef.pos,
+              s"A Scala.js-defined JS $obj cannot extend directly a native " +
```
cannot directly extend

gzm0 commented on compiler/src/main/scala/org/scalajs/core/compiler/PrepJSInterop.scala:
```diff
@@ -334,29 +375,42 @@ abstract class PrepJSInterop extends plugins.PluginComponent
+          val obj = if (sym.isTrait) "trait" else "class"
+          reporter.error(implDef.pos,
+              s"A Scala.js-defined JS $obj cannot extend directly a native " +
+              "JS trait. It must extend a JS class (native or not).")
```
The last sentence is misleading: This error condition can occur even if we directly extend a JS class (but also extend a native trait).

Also, a non-native trait can extend a non-native JS trait, can't it?

sjrd replied:
You're right. Copy/paste error from the error message in the previous `case`.

gzm0 commented on compiler/src/main/scala/org/scalajs/core/compiler/PrepJSInterop.scala:
```diff
@@ -365,39 +419,75 @@ abstract class PrepJSInterop extends plugins.PluginComponent
+        if (!isJSNative) OwnerKind.JSClass
+        else if (implDef.isInstanceOf[ModuleDef]) OwnerKind.JSNativeMod
+        else OwnerKind.JSNativeClass
+      }
```
From this I read that there is no such thing as a Scala.js defined JS module (which is reasonable IMO). However, I have not found the guard against that in this phase.

gzm0 replied:
Found it. Never mind.

gzm0 commented on compiler/src/main/scala/org/scalajs/core/compiler/PrepJSInterop.scala:
```diff
@@ -408,24 +498,41 @@ abstract class PrepJSInterop extends plugins.PluginComponent
-          case _ =>
-            reporter.error(tree.pos, "@JSBracketCall methods must have at "+
-                "least one non-repeated parameter")
+        if (enclosingOwner is OwnerKind.JSNative) {
```
I would prefer if you invert this check, so all the checks here use `JSClass`.

sjrd replied:
OK

gzm0 commented on compiler/src/main/scala/org/scalajs/core/compiler/PrepJSInterop.scala:
```diff
@@ -584,6 +684,12 @@ abstract class PrepJSInterop extends plugins.PluginComponent
 
   private def isJSDynamic(tree: Tree) = tree.tpe.typeSymbol == JSDynamicClass
 
+  /** Tests whether the symbol have `private` in any form, either `private`,
```
has

gzm0 commented on compiler/src/main/scala/org/scalajs/core/compiler/PrepJSInterop.scala:
```diff
@@ -437,15 +544,48 @@ abstract class PrepJSInterop extends plugins.PluginComponent
+        }
+
+        // Cannot override a private method (even with a protected/public one)
+        sym.allOverriddenSymbols.find(isPrivateMaybeWithin).foreach { overriddenSym =>
```
Are you 100% sure this gives only methods that are visible from `sym`?

sjrd replied:
Pretty sure. I'll add additional pos tests for that.

sjrd replied:
There are indeed problems with this. I will investigate later.

sjrd replied:
I found a much better idea: force `private[Scope]` methods to be `final` :-p And it is much easier to explain as a spec. Note that `private` and `private[this]` need not be final because they are implicitly so.

gzm0 replied:
`private` can be overridden in the companion (or in an inner class extending its outer class).

sjrd replied:
Nope, it can't, actually. I tried. It is visible and can be called from those places, but not overridden.

gzm0 commented on compiler/src/main/scala/org/scalajs/core/compiler/JSGlobalAddons.scala:
```diff
@@ -95,6 +95,12 @@ trait JSGlobalAddons extends JSDefinitions
+
+    def isExposed(sym: Symbol): Boolean =
+      sym.hasAnnotation(ExposedJSMemberAnnot)
+
```
I am assuming you need to do this, because in the backend you are unable to determine the orignal privateness reliably. If the privateness can be determined, we should not expose this in the frontend (or is this a binary compatibility concern)?

Further, I would refrain from using the name `register` here, since it is used differently in the same context (for per-run information). This should follow a equivalent path then adding and testing for the `@RawJSType` annotation. (Which is currently not abstracted but directly used in place).

I propose to either:

1. Remove and inline these two methods at usage site (or declare helpers in the relevant phases)
2. Change the name register, and change the `@RawJSType` logic to use the same flow.

I prefer 1, mainly because it allows us to have less stuff which is shared globally among phases.

sjrd replied:
OK I will do 1.

gzm0 commented on ir/src/main/scala/org/scalajs/core/ir/Trees.scala:
```diff
@@ -436,6 +436,122 @@ object Trees {
+  case class JSBracketSuperCall(cls: ClassType, receiver: Tree, method: Tree,
+      args: List[Tree])(implicit val pos: Position) extends Tree {
+    val tpe = AnyType
+  }
```
IIUC, this is basically a static call that first resolves to the super type of `cls`. Why not just have a `JSStaticBracketCall` (or however we call the static call thing of non-dynamic members)?

gzm0 replied:
Argh... I see why... You have no reliably way to refer to the parent of a JS class because it might be a raw JS class... OTOH, the compiler knows this, so it's maybe better to have the static call node and emit it conditionally based on what the super class is (in order to still allow for DCE).

sjrd replied:
The idea is never to emit different IR, when talking to another class' public or protected API, depending on whether that class is native or not. Basically, being native or not is an implementation detail of the class wrt other IR classes manipulating it.

gzm0 replied:
Nice! That makes total sense.

gzm0 replied:
Hmmm... I'm wondering whether `JSSuperBracketCall` / `JSSuperBracketSelect` wouldn't be a better name... Aligns nicely with `JSSuperConstructorCall`. WDYT?

sjrd replied:
OK, why not.

gzm0 commented on ir/src/main/scala/org/scalajs/core/ir/Trees.scala:
```diff
@@ -436,6 +436,122 @@ object Trees {
+   *
+   *  This node's semantics are to call the super constructor, then create
+   *  the fields of the object declared by the current class, initialized to
+   *  the zero of their type.
```
While I see that this is probably a good abstraction for "object initialization", I think the name is really poor. While it *does* call the super constructor, it is by far not everything it does...

How about: `JSBasicClassInitialization`?

sjrd replied:
I agree a better name would be nice, but I don't see how your suggestion solves the issue. It still fails to highlight the fact that the fields are created at that time. (or it fails to say that the super ctor is called, depending on how you interpret "Basic")

One solution might be to relax a bit the semantics of this node, in a way that could also allow us to take advantage of a possible hypothetical extension of JS in the far future that would create all fields of an object before constructors are called:

change the "This node's semantics" paragraph into:

> After the execution of this node, it is guaranteed that all fields declared in the current class have been created and initialized. Up to that point, accessing any field declared in this class (e.g., through an overridden method called from the super constructor) is undefined behavior.

This way, the node doesn't *provoke* the creation of the fields in its semantics. The field creation semantics is lifted as a guarantee of the structure of the constructor body itself (just like where `This()` can and cannot appear).

gzm0 commented on ir/src/main/scala/org/scalajs/core/ir/Trees.scala:
```diff
@@ -436,6 +436,122 @@ object Trees {
+   *  The presence of this node makes the class `Foo` instantiable, and
+   *  therefore reachable.
+   */
+  case class JSLoadConstructor(cls: ClassType)(
```
Why the name `load`? IIUC, it does not load anything at runtime (not like loading a module would). How about `JSClassConstructor`?

It has two advantages:
- Does not contain load
- Maybe it's a bit clearer that this is only for JS classes (which in the IR context mean Scala.js JS classes).

sjrd replied:
Actually, I will expand the semantics of this node to also be valid for a native JS class. This is necessary for native vs non-native to be an implementation detail. And that will make the "abuse" in desugaring a non-abuse.

And in that case, it might one day have to actually *load* stuff. We'll need some loading semantics when we have some support for `require`-style loading of native classes and objects.

gzm0 commented on compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala:
```diff
@@ -516,12 +599,43 @@ abstract class GenJSCode extends plugins.PluginComponent
+          else encodeFieldSym(f)
+
+        val irTpe = {
+          if (f.tpe.typeSymbol != CharClass || !isScalaJSDefinedJSClass(cd.symbol))
```
You are using `cd.symbol` here, above you use `currentClassSym`. Maybe have a `val` for the class symbol and assert the two are the same?

gzm0 commented:
Could we specifically disallow early initializers for Scala.js defined JS classes? IIUC they will not work anyways.

gzm0 commented on compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala:
```diff
@@ -627,13 +746,22 @@ abstract class GenJSCode extends plugins.PluginComponent
-              if (sym.isClassConstructor) {
-                js.MethodDef(static = false, methodIdent,
+              if (isJSClassConstructor) {
+                currentMethodInfoBuilder.setIsExported(true)
```
I don't understand this. Shouldn't these classes not be exported?

sjrd replied:
The class is not exported. It's `"constructor"` "method" is.

gzm0 replied:
Oh, I see. The constructor method needs to be exported since it needs to be added to the prototype of the class itself and can be retrieved via introspection in JavaScript. So if the class is instantiated, the ctor must be retained.

gzm0 commented on compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala:
```diff
@@ -733,6 +859,33 @@ abstract class GenJSCode extends plugins.PluginComponent
+          beforeSuper :::
+          afterSuper)(body.pos)
+    }
+
```
Ahem... This method seems not very safe... While I can see how this is necessary, shouldn't we be a bit more conservative? Maybe detect patterns that we allow to move (and document) and if we find something else fail?

We can do this all in asserts, so it can be turned off if it is hindering for performance...

sjrd replied:
As if we ever turn off asserts ^^

I initially wanted to do something like that, but stuff like `$outer` initialization is not very pretty. In general, anyway, I think we should move anything that scalac throws at us. It doesn't seem useful to prevent anything, actually. Or maybe just refuse to move `VarDef`s, since in theory they could declare a local value that would be used in a parameter to the super constructor (although I don't see how Scala source code could result in such a pattern).

gzm0 commented on compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala:
```diff
@@ -1457,6 +1661,22 @@ abstract class GenJSCode extends plugins.PluginComponent
+            reporter.error(pos,
+                "Implementation restriction: cannot call the super " +
+                "constructor of a Scala.js-defined JS class with :_*")
+          }
```
This will go away with better linker support, right?

sjrd replied:
Yes.

gzm0 commented on compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala:
```diff
@@ -3300,8 +3533,14 @@ abstract class GenJSCode extends plugins.PluginComponent
+      if (isScalaJSDefinedJSClass(sym)) {
+        val classFullName = encodeClassFullName(sym)
+        currentMethodInfoBuilder.addInstantiatedClass(classFullName)
+        js.JSLoadConstructor(jstpe.ClassType(classFullName))
```
Consider moving this part to `genPrimitiveJSClass` and assert here that we don't load a Scala.js defined class.

It will give us some safety against trying to load a Scala.js defined class as module.

(maybe also add asserts to the two methods above)...

gzm0 commented on compiler/src/main/scala/org/scalajs/core/compiler/GenJSExports.scala:
```diff
@@ -188,6 +195,44 @@ trait GenJSExports extends SubComponent { self: GenJSCode =>
+        val tpe = sym.tpe
+        if (!alts.exists(alt => tpe.matches(alt.tpe)))
+          alts ::= sym
+      }
```
I'm having the feeling that we already have code for this somewhere... Did you look for that?

sjrd replied:
Doesn't look like we have that anywhere else:
https://github.com/scala-js/scala-js/search?utf8=%E2%9C%93&q=isBridge
https://github.com/scala-js/scala-js/search?utf8=%E2%9C%93&q=matches

gzm0 replied:
Probably used to be code in the bridge generation in the good old 0.3 days :)

gzm0 commented on compiler/src/main/scala/org/scalajs/core/compiler/GenJSExports.scala:
```diff
@@ -188,6 +195,44 @@ trait GenJSExports extends SubComponent { self: GenJSCode =>
+      } else if (isProp && alts.exists(_.owner != classSym)) {
+        reporter.error(alts.head.pos,
+            "Illegal override/overload of inherited getter/setter for " +
+            s"${classSym.fullName}::$name. All alternatives must be covered.")
```
I do not understand this error condition. Would you mind elaborating?

sjrd replied:
Ah, this is in fact an implementation restriction. It comes from the inability to call a super getter or super setter. But in fact ES 6 can do it, and [so can Babel](https://github.com/google/traceur-compiler/blob/67180b377559adeb330260e95b878afb99e826af/src/runtime/classes.js#L29-L64). So I'll fix that.

gzm0 commented:
That is all for the changes to the compiler code and IR (w/o IR checker) in "Essentials of Scala.js-defined JS classes". Will continue later

gzm0 commented on tools/shared/src/main/scala/org/scalajs/core/tools/optimizer/Linker.scala:
```diff
@@ -224,19 +224,16 @@ final class Linker(semantics: Semantics, outputMode: OutputMode,
-            analyzerInfo.methodInfos(m.name.name).isReachable) {
-          if (m.body == EmptyTree)
+      case m: MethodDef =>
+        if (analyzerInfo.methodInfos(m.name.name).isReachable) {
```
I'm assuming `isReachable` already accounts for subclass instantiation and exports. Could you confirm this?

sjrd replied:
Yes. `isReachable` means that this particular implementation of the method is reachable. It must be kept.

gzm0 commented on tools/shared/src/main/scala/org/scalajs/core/tools/optimizer/IRChecker.scala:
```diff
@@ -163,6 +171,11 @@ class IRChecker(unit: LinkingUnit, logger: Logger) {
 
+    if (classDef.kind == ClassKind.JSClass && !static) {
+      reportError(s"Non exported instance method $name is illegal in JS class")
+      return // things would too badly otherwise
```
Typo in comment

gzm0 commented on tools/shared/src/main/scala/org/scalajs/core/tools/optimizer/IRChecker.scala:
```diff
@@ -233,14 +246,63 @@ class IRChecker(unit: LinkingUnit, logger: Logger) {
-    typecheckExpect(body, bodyEnv, resultType)
+  def checkJSClassConstructor(methodDef: MethodDef,
+      classDef: LinkedClass): Unit = {
+    val MethodDef(static, StringLiteral(name), params, resultType, body) = methodDef
```
You don't use the name here. Either assert is is `== "constructor"`, or ignore it.

gzm0 commented:
Note to self:
Stuff to verify there are tests for:
- Overriding of private members (neg test)
- Private members with same name in subclass (pos test)

gzm0 commented on ir/src/main/scala/org/scalajs/core/ir/ClassKind.scala:
```diff
@@ -26,6 +26,7 @@ object ClassKind {
   case object Interface extends ClassKind
   case object RawJSType extends ClassKind
   case object HijackedClass extends ClassKind
+  case object JSClass extends ClassKind
```
I think adding `isAnyClass` to `ClassKind` might be a good idea, given how many times `x.isClass || x == ClassKind.JSClass` occurs in the codebase.

sjrd replied:
OK

gzm0 commented on tools/shared/src/main/scala/org/scalajs/core/tools/javascript/JSDesugaring.scala:
```diff
@@ -420,6 +394,60 @@ object JSDesugaring {
+
+            val superCtorCall = {
+              val superCtor = genRawJSClassConstructor(
+                  getSuperClassOfJSClass(linkedClass))
```
You only need this in the ES5 branch below. You should move it there.

gzm0 commented on tools/shared/src/main/scala/org/scalajs/core/tools/javascript/JSDesugaring.scala:
```diff
@@ -420,6 +394,60 @@ object JSDesugaring {
+              val descriptors = js.ObjectConstr(List(
+                  transformedName -> descriptor))
+              js.Apply(defineProperties, List(js.This(), descriptors))
+            }
```
Why not simple assings?

gzm0 replied:
Also, why not batch into a single call to `defineProperties`?

sjrd replied:
Simple assigns don't allow a `val` (field) to override an inherited *def* (getter).

No batch because, in theory, es 5 does not specify that the order of the fields is kept when iterating over them, which means that the fields could be created in different orders on different objects, which is bad for performance.

gzm0 replied:
Ok. Makes sense. Btw. I am wondering why we don't use `writable` to implement actual vals (this would be the time to do it).

sjrd replied:
We don't use `writable: false` because we actually do reassign it later (in the initialization coming from the constructor). Moreover, creating a property with anything but `{ configurable: true, enumerable: true, writable: true }` causes current VMs to switch to hash table mode, instead of the optimized hidden class representation for the object.

gzm0 commented on tools/shared/src/main/scala/org/scalajs/core/tools/javascript/JSDesugaring.scala:
```diff
@@ -1199,6 +1228,21 @@ object JSDesugaring {
+          val superClass = getSuperClassOfJSClass(
+              classEmitter.linkedClassByName(cls.className))
+          val superCtor =
+            JSLoadConstructor(ClassType(superClass.encodedName)) // abuse
```
You don't need to abuse. Just call `genRawJSClassConstructor(superClass)`

gzm0 replied:
Or even better: `encodeClassVar(superClass.name)`.

sjrd replied:
Doesn't work. I need IR trees here, not a js tree.

gzm0 replied:
Ah... OK. Makes sense...

gzm0 commented on tools/shared/src/main/scala/org/scalajs/core/tools/javascript/JSDesugaring.scala:
```diff
@@ -1199,6 +1228,21 @@ object JSDesugaring {
+                StringLiteral("call"),
+                receiver :: args)
+          }
+
```
Shouldn't this be conditionally translated to a super call in ES6?

sjrd replied:
It could in the future, but it is by no means a requirement.

gzm0 replied:
Ok.

gzm0 commented on tools/shared/src/main/scala/org/scalajs/core/tools/javascript/JSDesugaring.scala:
```diff
@@ -1199,6 +1228,21 @@ object JSDesugaring {
+          redo {
+            JSBracketMethodApply(
+                JSBracketSelect(
+                    JSBracketSelect(superCtor, StringLiteral("prototype")),
```
Bracket select of prototype? Are you sure?

sjrd replied:
It doesn't really matter. I used `JSBracketSelect` simply not to use `JSDotSelect`. I think in general `JSDotSelect` should be used only for things we can rename. And we cannot rename `prototype`.

gzm0 commented on tools/shared/src/main/scala/org/scalajs/core/tools/javascript/ScalaJSClassEmitter.scala:
```diff
@@ -194,20 +236,23 @@ final class ScalaJSClassEmitter(semantics: Semantics, outputMode: OutputMode,
-    }
+    val inheritableCtorDef =
+      if (isJSClass) js.Skip()
+      else makeInheritableCtorDef(typeVar)
```
What's the rationale to not always emit an inheritable ctor for JS Classes just like for normal classes?

sjrd replied:
Because we'll never use it. 
We don't use it because the subclass creates its own inheritable constructor, in case its superclass is native. This schema is actually more typical for declaring classes in js. The one used for Scala classes is a tiny space optimization.

gzm0 replied:
Hmmm... OK. So you just don't want to do the space optimization here? Since we are at link time, an extending class can just check the type of its superclass and emit an inheritable ctor or not based on that.

gzm0 replied:
I guess we can leave it like that, since it keeps Scala defined JS classes and native JS classes closer together.

gzm0 commented on tools/shared/src/main/scala/org/scalajs/core/tools/javascript/ScalaJSClassEmitter.scala:
```diff
@@ -216,27 +261,52 @@ final class ScalaJSClassEmitter(semantics: Semantics, outputMode: OutputMode,
+    } getOrElse {
+      throw new IllegalArgumentException(
+          s"${tree.encodedName} does not have an exported constructor")
+    }
```
I do not understand this. How is an exported and a not-exported ctor of a JS class different?

gzm0 replied:
Oh, I see. This is not actually exported, the JS ctor just gets classified in the `exportedMembers` field of the `LinkedClass`. How can we (on an IR level) distinguish an exported JS class from a non-exported one (i.e. what is the plan to do so)?

Shouldn't we stick with the `init___` convention and consider the fact that in ES6 the constructor is called "constructor" an implementation detail? It seems odd to have multiple magic names if there is no need to do so...

sjrd replied:
A class will be exported if it contains a `ExportJSClassDef` in its defs (similar to `ExportModuleDef`). The constructor being exported is normal assuming the class is reachable (by any means). It is completely orthogonal to the class being exported or not.

When exporting a js class, we do not export its constructors. We export the class itself (unlike scala classes)

gzm0 replied:
Ok. This makes sense now.

gzm0 commented on tools/shared/src/main/scala/org/scalajs/core/tools/javascript/ScalaJSClassEmitter.scala:
```diff
@@ -216,27 +261,52 @@ final class ScalaJSClassEmitter(semantics: Semantics, outputMode: OutputMode,
+    val MethodDef(_, _, params, _, body) = ctor
+
+    desugarToFunction(this, tree.encodedName,
+        params, body, isStat = true)
```
Consider moving this block into the `collectFirst`, so you can match on the `MethodDef` once (and you won't need the guards anymore).

gzm0 commented on tools/shared/src/main/scala/org/scalajs/core/tools/javascript/JSDesugaring.scala:
```diff
@@ -107,59 +109,29 @@ import java.io.StringWriter
-        throw new DesugarException(tree, cause)
-    }
-  }
-
```
Why did you remove `desugarJavaScript`? I can't see an inherent reason that we can't do it anymore (of course only given the additional info we need now).

sjrd replied:
The lack of that information must then be threaded through the desugar, and this complicates the code. I tried and it's not pretty. Since it doesn't work anymore in that case anyway, it should be removed (better to break at link time than silently fail at runtime). Besides, it's a very low exposure api.

gzm0 replied:
I meant to add the following to `JSDesugaring`:

```scala
def desugarJavaScript(classEmitter: ScalaJSClassEmitter, enclosingClassName: String, body: Tree, isStat: Boolean): js.Tree = {
  val desugar = new JSDesugar(classEmitter, enclosingClassName, None)
  if (isStat)
    desugar.transformStat(body)(Env.empty)
  else
    desugar.transformExpr(body)(Env.empty)
}
```

and use it instead of the `fakeFunction` below.

Or am I completely missing something?

sjrd replied:
Oh, OK.

sjrd commented:
Partially updated. There is one main change: `JSLoadConstructor` allows native JS classes. This actually simplifies several things, and it gave me an easy way to fix the non-optimality regarding the dce of class data of a native JS class extended by a Scala.js-defined JS class.

Will do the rest later.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/2043/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/2043/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2647/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2647/</a>
Test PASSed.

sjrd commented:
> Could we specifically disallow early initializers for Scala.js defined JS classes? IIUC they will not work anyways.

I haven't considered that. I guess allowing them would be misleading. I'll see whether we can identify and disallow them as an early error.

sjrd commented:
Updated.

TODO:

* Call super getter and super setter
* `allOverriddenMethods` stuff -> check how refchecks does it
* Disallow early initializers
* Refuse to move `VarDef`s after the `JSSuperConstructorCall`

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/2044/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/2044/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2649/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2649/</a>
Test FAILed.

sjrd commented:
> Disallow early initializers

Unfortunately, already after `typer` we have lost the information that something was an early initializer, as opposed to a `val/var` parameter of the constructor. So disallowing this is not implementable.

gzm0 commented:
This is still WIP, right? So no need to review just yet.

sjrd commented:
Yes, still wip at this point.

sjrd commented:
Updated. Good to review again.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/2064/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/2064/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2676/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2676/</a>
Test FAILed.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/2065/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/2065/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2677/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2677/</a>
Test FAILed.

sjrd commented:
Updated. Fixed checksizes and fixed the problems about default parameters.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/2068/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/2068/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2682/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2682/</a>
Test PASSed.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/2069/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/2069/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2683/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2683/</a>
Test FAILed.

sjrd commented:
retest this please

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/2076/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/2076/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2689/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2689/</a>
Test FAILed.

gzm0 commented:
LGTM until ec28d7a96c31e07d04885770dd79730087047ed2

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/2100/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/2100/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2716/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2716/</a>
Test FAILed.

gzm0 commented on compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala:
```diff
@@ -510,18 +593,53 @@ abstract class GenJSCode extends plugins.PluginComponent
+            toIRType(f.tpe)
+          else
+            jstpe.ClassType(ir.Definitions.BoxedCharacterClass)
+        }
```
I do not understand this logic. IIUC, we'll have to store exposed fields in boxed form. However, this doesn't handle value classes.

Also, why do we box private fields?

sjrd replied:
> However, this doesn't handle value classes.

Indeed, [it doesn't](https://github.com/scala-js/scala-js/pull/1809/files#diff-6e8c360252085a4ec6f673609bd75631R238). I have not found a way to support this reliably so far. Fortunately this happens only for fields whose static type is a value class, and that are not initialized, i.e., they have `= _`. I guess we could forbid this in PrepJSInterop, now that I think about it.

> Also, why do we box private fields?

Consistency in the implementation. Private fields are rare enough and boxing cheap enough (basically no-op except for Char) that it is not worth complicating the implementation to specifically de-box private fields.

gzm0 replied:
> I have not found a way to support this reliably so far.

What is wrong with erasing the field to `Any` in `jsinterop` and let scalac handle the rest?

In any case, please comment about this here (if you intend to not change it).

sjrd replied:
> What is wrong with erasing the field to Any in jsinterop and let scalac handle the rest?

The same as if we erased it right here: an *uninitialized* `var` (with `= _`) will contain the zero of a primitive type (e.g., `null`), but accesses to the field will read it as `obj.field.asInstanceOf[ValueClass].underlyingField` which will NPE at best.

This is because scalac (erroneously IMO) doesn't treat `null` as the 0 of the underlying type when "unboxing" a ValueClass.

Since the field is directly accessible by JS, we cannot control reads of the field to fix up `null` at the time of reading. And hence a `null` can leak out, be passed around, and eventually arrive at an unboxing operation which will NPE (and it shouldn't have, since it should have be an actually boxed zero instead of `null`).

gzm0 replied:
Scalac has the same issue:

```scala
scala> class Foo(val x: Int) extends AnyVal
defined class Foo

scala> class A[T] { var x: T = _ }
defined class A

scala> new A[Foo]
res1: A[Foo] = A@39d186

scala> res1.x.x
java.lang.NullPointerException
  ... 43 elided
```

But currently, we are doing it wrong differently (value class fields never work, right? Or did I get this wrong?). We should probably replicate Scalac's behavior...

sjrd replied:
scalac has the issue if the field is of a *generic type* that happens to be instantiated to a value class. We inherit from this issue, of course.

But for Scala.js-defined JS classes, we also have the issue if the field has a value class type (monomorphically). scalac does not have that issue.

I will add a comment there explaining that there is some fundamental limitation, here.

gzm0 replied:
> But for Scala.js-defined JS classes, we also have the issue if the field has a value class type (monomorphically). scalac does not have that issue.

That is fine. Per specification, the JS world works just like the boxed world in Scala. Therefore, the comparison with a generic type that happens to be of a value class type is accurate.

IMO your comment there & the test is fine.

gzm0 commented on compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala:
```diff
@@ -1361,20 +1573,27 @@ abstract class GenJSCode extends plugins.PluginComponent
 
+        case _ if isRawJSDefaultParam =>
+          js.UndefinedParam()(toIRType(sym.tpe.resultType))
+
```
Hahaha, funny bug :) It's much cleaner here anyways IMO.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/2101/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/2101/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2717/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2717/</a>
Test FAILed.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/2103/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/2103/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2719/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2719/</a>
Test FAILed.

sjrd commented:
Need to update checksizes after rebase.

sjrd commented:
Rebased and fixed checksizes.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/2115/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/2115/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2734/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2734/</a>
Test FAILed.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/2118/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/2118/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2737/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2737/</a>
Test PASSed.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/2119/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/2119/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2738/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2738/</a>
Test PASSed.

sjrd commented:
Rebased after the 2.12.0-M2 PR was merged. Note the new Workaround commit which disables MiMa on 2.12.0-M2.

gzm0 commented on compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala:
```diff
@@ -3092,9 +3324,55 @@ abstract class GenJSCode extends plugins.PluginComponent
+
+    private def genJSSuperCall(tree: Apply, isStat: Boolean): js.Tree = {
+      implicit val pos = tree.pos
+      val Apply(fun @ Select(sup @ Super(_, mix), _), args) = tree
```
You don't use `mix` in the body.

gzm0 commented on compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala:
```diff
@@ -3092,9 +3324,55 @@ abstract class GenJSCode extends plugins.PluginComponent
+      lazy val genJSArgs = genPrimitiveJSArgs(sym, args)
+
+      if (sym.owner == ObjectClass) {
+        // Normal call anyway
```
Add a comment that this is a trait ctor case.

sjrd replied:
Er no, it isn't.

That's a call like `super.hashCode()` which resolves to Object's `hashCode` implementation. But we can't really call `$c_jl_Object.prototype.hashCode__I.call(this)` here because `this` is *not* an instance of `jl.Object` at this point. However, it is perfectly valid to generate `ApplyMethod(this, "hashCode")` because, `this` being of type `AnyType`, this is rerouted to the helper `$objectHashCode`, which returns 42/identity hash code for the argument.

gzm0 replied:
Ah :) Somehow I thought this is only about super ctors... :)

gzm0 commented on compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala:
```diff
@@ -3092,9 +3324,55 @@ abstract class GenJSCode extends plugins.PluginComponent
+        genApplyJSClassMethod(genReceiver, sym, genScalaArgs)
+      } else {
+        genJSCallGeneric(sym, genReceiver, genJSArgs, isStat,
+            superIn = Some(currentClassSym))
```
When do we hit this case?

gzm0 replied:
Also, doesn't this hit the most simple `genCall` case only in `genJSCallGeneric`? And you could just replace this with:

```scala
js.JSBracketSuperCall(
 jstpe.ClassType(encodeClassFullName(superInSym)),
 receiver, methodName, args)
```

sjrd replied:
> When do we hit this case?

Most of the time. Almost every time you call `super` in a Scala.js-defined JS class.
```scala
class NativeParent extends js.Object {
  def someGetter: Int = js.native
  def foo(x: Int): Int = js.native
  def apply(x: String): String = js.native
}

@ScalaJSDefined
class ScalaJSDefinedChild extends NativeParent {
  override def someGetter: Int = super.someGetter + 3
  override def foo(x: Int): Int = super.foo(x) + 1
  def foobar(x: Int): Int = super.foo(x) * 2
  def bar(x: String): String = super.apply(x)
}
```
All lines in `ScalaJSDefinedChild` end up in this case. Also applies if `NativeParent` is in fact non native. The fourth one hits the special case for `apply` in `genJSCallGeneric`, which will cause a compile-error. You can also call `super` on bracket-select and bracket-call symbols. `genJSCallGeneric` captures all of that.

gzm0 replied:
Yeah, makes absolute sense... I thought this is super ctors only... :S

gzm0 commented on compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala:
```diff
@@ -3102,51 +3380,81 @@ abstract class GenJSCode extends plugins.PluginComponent
+              js.JSBracketSuperSelect(
+                  jstpe.ClassType(encodeClassFullName(superInSym)),
+                  receiver, propName)
+            }
```
```scala
val lhs = genSelectGet(propName)
```
?

sjrd replied:
Ah ah, true, they *do* the same thing. Although it's semantically completely unrelated, which I guess is why I did not "see" the duplication. I'll add a third `def` to delegate to in `genSelect{Get,Set}`.

gzm0 commented on compiler/src/main/scala/org/scalajs/core/compiler/JSEncoding.scala:
```diff
@@ -61,7 +61,7 @@ trait JSEncoding extends SubComponent { self: GenJSCode =>
     }
     usedLocalNames += longName
-    longName
+    mangleJSName(longName)
```
Shouldn't we drop the `isKeywordOrReserved` in the loop in this case?

sjrd replied:
I almost said "yes, we can do that". But he problem is the `OrReserved`, i.e., `"arguments"`, `"eval"` and `ScalaJSEnvironmentName == "ScalaJS"`. These are not taken into account by `mangleJSName`. Note that we cannot "fix" `mangleJSName` because it defines how public names such as methods are mangled, so changing the algorithm of `mangleJSName` breaks binary compatibility.

I could change `isKeywordOrReserved(longName)` into `isReserved(longName)`, though.

gzm0 replied:
Oh... Didn't even realize the reserved... This is actually a bug in `mangleJSName`, no? We should probably file this for 1.0.

In this case, maybe better we don't change this...

sjrd replied:
No, it is not a bug, because all other uses of `mangleJSName` are accompanied by some suffix. For example for methods you have a suffix `__xyz`. For fields you have a suffix `$f` or `$1`. Moreover, all of these uses are for identifiers in the *property* namespace of JS (for after a `.`), where reserved identifiers can be legally used anyway (but not keywords). This is why only `freshLocalName` handles reserved identifiers specially.

gzm0 replied:
Ah OK. I guess in that case changing to `isReserved` is probably the right thing to do.

gzm0 commented on compiler/src/main/scala/org/scalajs/core/compiler/JSEncoding.scala:
```diff
@@ -218,7 +218,12 @@ trait JSEncoding extends SubComponent { self: GenJSCode =>
+    val paramTypeNames =
+      if (!isScalaJSDefinedJSClass(sym.owner)) paramTypeNames0
+      else internalName(sym.owner.toTypeConstructor) :: paramTypeNames0
+
```
This totally overlaps with how the runtime classes are handled (same code is below again). Consider generalizing the `inRTClass` flag and not checking for `isScalaJSDefinedJSClass` in this file. It seems off...

sjrd replied:
Hum, good catch. Generalizing `inRTClass` seems risky, but I can definitely factor out the ways the first parameter type is prepended to the signature.

gzm0 replied:
OK.

gzm0 commented on compiler/src/main/scala/org/scalajs/core/compiler/PrepJSInterop.scala:
```diff
@@ -365,39 +423,75 @@ abstract class PrepJSInterop extends plugins.PluginComponent
+          }
+        }
+
+        // Register exposed symbols in Scala.js-defined classes
```
Outdated comment

gzm0 commented on compiler/src/main/scala/org/scalajs/core/compiler/PrepJSInterop.scala:
```diff
@@ -408,24 +502,41 @@ abstract class PrepJSInterop extends plugins.PluginComponent
+            case (param :: _) :: _ if !isScalaRepeatedParamType(param.tpe) =>
+              // ok
+            case _ =>
+              reporter.error(tree.pos, "@JSBracketCall methods must have at "+
```
missing space before +

gzm0 commented on compiler/src/main/scala/org/scalajs/core/compiler/PrepJSInterop.scala:
```diff
@@ -437,15 +548,48 @@ abstract class PrepJSInterop extends plugins.PluginComponent
+            !sym.isFinal) {
+          reporter.error(tree.pos,
+              s"The qualified private method ${sym.nameString} in a " +
+              "Scala.js-defined JS class must be final")
```
That is a strange message. To me it means, that only the method with this name must be final... Others are OK. IMO there is no need to put the name in the method... The error will indicate the right location.

sjrd replied:
ok

gzm0 commented on compiler/src/test/scala/org/scalajs/core/compiler/test/ScalaJSDefinedTest.scala:
```diff
@@ -0,0 +1,451 @@
+      |newSource1.scala:8: error: The qualified private method foo in a Scala.js-defined JS class must be final
+      |        private[Enclosing] def foo(i: Int): Int = i
+      |                               ^
+    """
```
Could you add such a test for a `val` and a `var`, too?

sjrd replied:
ok

gzm0 commented on ir/src/main/scala/org/scalajs/core/ir/Infos.scala:
```diff
@@ -47,6 +47,9 @@ object Infos {
       val staticMethodsCalled: Map[String, List[String]],
+      /** For a Scala class, it is instantiated with a `New`; for a JS class,
+       *  its constructor is accessed with a `JSLoadConstructor`.
+       */
```
I'm not sure this is the best place to put this... But I don't know a better place either... :-/

gzm0 commented on ir/src/main/scala/org/scalajs/core/ir/Infos.scala:
```diff
@@ -356,6 +359,9 @@ object Infos {
 
+        case JSLoadConstructor(cls) =>
+          builder.addInstantiatedClass(cls.className)
+
```
Hmm... This is actually an over approximation... If the `JSLoadConstructor` is inside an `isInstanceOf`, it does not make the class instantiated...

gzm0 replied:
This is actually funny :P
In the case where we use a class only in an `isInstanceOf`, if we don't use this over approximation, the unoptimized code will fail, but optimized code will be able to elide the check entirely...

I guess leaving it for now is fine then :)

gzm0 commented on test-suite/src/test/scala/org/scalajs/testsuite/jsinterop/ScalaJSDefinedTest.scala:
```diff
@@ -0,0 +1,978 @@
+      expect(dyn.bar(4, 5)).toEqual(9)
+      expect(dyn.bar(4)).toEqual(5)
+      expect(dyn.foobar(3)).toEqual(4)
+    }
```
Could you add a test with dependent default params (`def foo(x: Int)(y: Int = x + 1)`) please?

sjrd replied:
ok

gzm0 commented on tools/shared/src/main/scala/org/scalajs/core/tools/optimizer/Analyzer.scala:
```diff
@@ -349,19 +352,34 @@ final class Analyzer(semantics: Semantics, outputMode: OutputMode,
+          for (methodInfo <- methodInfos.values) {
+            if (methodInfo.isExported)
+              methodInfo.reach(this)(FromExports)
+          }
```
Why is this not done in the same place than for normal classes? IIUC, the export reachability for normal classes is actually too aggressive. It reaches even if the class is not instantiated...

sjrd replied:
Exports on normal classes take overriding into account. With:
```scala
class Foo {
  @JSExport def foobar(): Int = 42
}
class Bar extends Foo {
  @JSExport override def foobar(): Int = 3
}
new Bar
```
you'll have that `Foo["foobar"]` can be dce'ed away, because `Foo` is not instantiated. Only `Bar["foobar"]` is reachable. This is implemented through `callMethod` which actually *delays* the actual call until the class is `instantiated()`. We are allowed to do this because the semantics of Scala classes and exports do not guarantee anything about what will be present in the `prototype`s.

For JS classes, however, it is necessary to expose everything as soon as the constructor is reachable, because the classes could be extended by JS classes, or otherwise reflected upon using the JS prototype.

gzm0 replied:
Ah, makes sense. I was confused by `callMethod` and `delayedCallMethod`...

gzm0 commented on tools/shared/src/main/scala/org/scalajs/core/tools/optimizer/IRChecker.scala:
```diff
@@ -233,14 +246,63 @@ class IRChecker(unit: LinkingUnit, logger: Logger) {
+    }
+
+    for (arg <- superCall.args)
+      typecheckExprOrSpread(arg, preparedEnv)
```
I thought we don't allow spread here yet.

sjrd replied:
There are three things involved here:

1. The IR spec allows it.
2. The desugar has a known bug wrt. the spec, and will crash.
3. As a workaround, the compiler prevents the generation of spread in this case to anticipate the desugar crash, and turns it into a nice compile error.

But basically the `IRChecker` is the spec, so it allows it.

Does that make sense?

gzm0 replied:
Hahaha, OK. But let's avoid this in a release version, ok?

sjrd replied:
Unfortunately, it is very unlikely that we *can* avoid this. As mentioned in #1813, we cannot fix the desugar until io.js supports spread calls. This bug will have to stay open until then.

gzm0 replied:
Hmmm... In this case I don't think we should prevent these for ES5, just because ES6 support is not mature enough yet. I'd rather have the desugar fail in ES6 mode if something like this occurs...

sjrd replied:
That would set a precedent for something we support when targeting ES 5, and that we don't support when targeting ES 6 :-s It would also be virtually impossible to test; we'd need to have a file that we include in the compilation only when targeting ES 5. (no, a Jasmine flag does not work here, because linking would crash before the flag is tested at runtime)

gzm0 replied:
Can io.js parse the file? If yes, let it crash at runtime (there, the flag will work, right?). We say we emit ES6. If the VM does not support ES6, it's not our job.

In other words: I don't agree on the precedent argument, since the culprit here is the VMs lacking support, but I see that testability is an issue...

sjrd replied:
No, I don't think it can parse the file.

But I could test locally once that it works on ES 5, then comment out the test until we can test it regularly with everything else on ES 6?

gzm0 commented on tools/shared/src/main/scala/org/scalajs/core/tools/optimizer/IRChecker.scala:
```diff
@@ -680,6 +746,30 @@ class IRChecker(unit: LinkingUnit, logger: Logger) {
 
+      case JSBracketSuperSelect(cls, qualifier, item) =>
+        if (lookupClass(cls).kind != ClassKind.JSClass)
+          reportError(s"JS class type expected by $cls found")
```
but

gzm0 commented on tools/shared/src/main/scala/org/scalajs/core/tools/optimizer/IRChecker.scala:
```diff
@@ -680,6 +746,30 @@ class IRChecker(unit: LinkingUnit, logger: Logger) {
+
+      case JSBracketSuperCall(cls, receiver, method, args) =>
+        if (lookupClass(cls).kind != ClassKind.JSClass)
+          reportError(s"JS class type expected by $cls found")
```
but

gzm0 commented on tools/shared/src/main/scala/org/scalajs/core/tools/optimizer/IRChecker.scala:
```diff
@@ -680,6 +746,30 @@ class IRChecker(unit: LinkingUnit, logger: Logger) {
+          case _                   => false
+        }
+        if (!valid)
+          reportError(s"JS class type expected by $cls found")
```
but

gzm0 commented on tools/shared/src/main/scala/org/scalajs/core/tools/optimizer/Linker.scala:
```diff
@@ -224,19 +224,16 @@ final class Linker(semantics: Semantics, outputMode: OutputMode,
-      case m: MethodDef if m.name.isInstanceOf[StringLiteral] =>
-        if (analyzerInfo.isAnySubclassInstantiated)
-          exportedMembers += linkedMethod(m)
-
```
Ah, ah :) this is were we made the exports on normal classes non-aggressive. Your solution is definitely better, but you'll have to adapt the analyzer (where I commented).

gzm0 replied:
So this is obsolete as well...

gzm0 commented:
That's all until eda1c38ac143d0ce7abbec5667ac4139804f1487

gzm0 commented on compiler/src/test/scala/org/scalajs/core/compiler/test/ScalaJSDefinedTest.scala:
```diff
@@ -0,0 +1,451 @@
+      |      def foo(x: Int): Int = x + 1
+      |          ^
+    """
+  }
```
Please also add a neg test for private members in traits

sjrd replied:
ok

gzm0 replied:
Did you address this?

gzm0 replied:
Ah, yes, I saw it. The error message is a bit convoluted, but that's fine. No need to complicate compiler code. Who puts private methods in traits you cannot put implementations in anyways...

gzm0 commented:
That's all.

sjrd commented:
Updated. I should have addressed all the comments, except the fixes to super constructor call with spread. What would like me to do: fix directly within "Essentials of ..."; fix as an additional commit referencing #1813; or fix in a separate PR for #1813?

sjrd commented:
> What would like me to do: fix directly within "Essentials of ..."; fix as an additional commit referencing #1813; or fix in a separate PR for #1813?

Well, actually, we lack the infrastructure in desugar to emit spread calls, at the moment. So fixing it requires substantial changes which should definitely go in a different PR anyway.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/2120/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/2120/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2741/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2741/</a>
Test PASSed.

gzm0 commented:
> Well, actually, we lack the infrastructure in desugar to emit spread calls, at the moment. So fixing it requires substantial changes which should definitely go in a different PR anyway.

Completely fine with that.

gzm0 commented:
LGTM

sjrd commented:
Yeepee! :-)
Thanks a lot for that review.

gzm0 commented:
![Brace Yourself - The followup PRs are coming](http://cdn.meme.am/instances2/500x/1525318.jpg)

sjrd commented:
You have no idea :-p
But unlike the introduction of non-boxing after -- what was it? -- 0.4.4, I haven't them ready yet ;-)

gzm0 commented:
I don't know. But it was terrible :)

After 2000 LOC PR:
Me: "LGTM, phew done"
You: "Ohhhhhhh, btw, I had these nice 5 1000 LOC PRs each depending on the one we just merged"
Me: :sob:
